src/vmm/benches/queue.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Benchmarking cases:
+//   * `Queue.pop`
+//   * `Queue.add_used`
+//   * `DescriptorChain.next_descriptor`
+
+use std::num::Wrapping;
+
+use criterion::{criterion_group, criterion_main, Criterion};
+use vm_memory::GuestAddress;
+use vmm::devices::virtio::queue::{VIRTQ_DESC_F_NEXT, VIRTQ_DESC_F_WRITE};
+use vmm::devices::virtio::test_utils::VirtQueue;
+use vmm::utilities::test_utils::single_region_mem;
+
+/// Create one chain with n descriptors
+/// Descriptor buffers will leave with offest of 2048 bytes
+/// to leave some room for queue objects themselfs.
+/// Even though we don't really care about size of descriptors
+/// the maximum size of the packet is 64K and we are
+/// building a chain that has this length. For this
+/// we need each descriptor to be 64K / n size.
+fn set_dtable_one_chain(rxq: &VirtQueue, n: usize) {
+    for i in 0..n {
+        rxq.dtable[i].set(
+            (2048 + (65562 / n) * i) as u64,
+            (65562 / n) as u32,
+            VIRTQ_DESC_F_WRITE | VIRTQ_DESC_F_NEXT,
+            (i + 1) as u16,
+        );
+    }
+    rxq.dtable[n - 1].flags.set(VIRTQ_DESC_F_WRITE);
+    rxq.dtable[n - 1].next.set(0);
+    rxq.avail.ring[0].set(0);
+    rxq.avail.idx.set(n as u16);
+}
+
+/// Create n chains with 1 descriptors each
+/// Descriptor buffers will leave with offest of 2048 bytes
+/// to leave some room for queue objects themselfs.
+/// Even though we don't really care about size of descriptors
+/// the maximum size of the packet is 64K and we are
+/// building a set of descriptors with this total size.
+/// For this we need each descriptor to be 64K / n size.
+fn set_dtable_many_chains(rxq: &VirtQueue, n: usize) {
+    for i in 0..n {
+        rxq.dtable[i].set(
+            (2048 + (65562 / n) * i) as u64,
+            (65562 / n) as u32,
+            VIRTQ_DESC_F_WRITE,
+            0,
+        );
+        rxq.avail.ring[i].set(i as u16);
+    }
+    rxq.avail.idx.set(n as u16);
+}
```

Comment:
If I understand this correctly, you are saying that a single packet will be distributed across different `DescriptorChain` objects? If that's what you mean, I don't think this is possible from the VirtIO spec. Even if it was, TAP device doesn't work like that. Every write in the TAP device should be a single Ethernet frame. Also every read returns exactly one ethernet frame.

Reply:
This is just to test `Queue::pop` method, so we don't need to be super realistic here.
I guess comments about making the total len of the chain being 64K might be confusing, because technically it does not matter as benches do not read/write to these buffers.
What is your suggestion here?

Reply:
I think this scenario is not realistic. We will never find a queue filled with `DescriptorChain`s of length one. So we should not benchmark for this scenario.

Reply:
Ok, my bad. I just understood what you mean. You're not parsing the chain after pop. Then it's fine, but I'd drop the logic for splitting 64K across descriptors. It's confusing. Just create identical Descriptors with fixed sizes.

Reply:
I am benching the `Queue::pop` method with chains created with this function. The length of each chain is irrelevant here, as we don't do anything with it. So creating N chains of length 1 is fine. Creating chains of lther lengths will require more complex logic and not worth it in this case.

Reply:
Probably I wasn't clear. That's exactly what I meant. Just fill up the queue with the same Descriptor. Choose a single size and that's it.

Reply:
Ok, done